### PR TITLE
Create pyright.yml

### PR DESCRIPTION
**Pull request by nelson-lojo**
Add a github action that makes sure type checking is satsfied for PRs into main. (this should serve as a stop-gap in lieu of tests)

### CODE REVIEW

**Comment by nelson-lojo** [comment]
image

**Comment by nelson-lojo** [comment]
The $ when trying to expand the conda directory is getting escaped sometime after variable expansion is done
image

File: src/__main__.py [hunk 1]
```diff
@@ -14,19 +14,19 @@ def nest_yaml(tag: str, content: str, indent_size:int=4) -> str:
 
 def log_yaml(full_yaml: str) -> None:
     # save locally
-    local_dir_path = f"models/{os.path.basename(os.path.dirname(wandb.run.dir)).replace('run-', '')}"
+    local_dir_path = f"models/{os.path.basename(os.path.dirname(wandb.run.dir)).replace('run-', '')}" # pyright: ignore [reportOptionalMemberAccess]
```
Comment (nelson-lojo) [inline src/__main__.py:17]: Adding an ignore comment *every* time we use `wandb.run.dir` seems verbose ... is there a way to ensure that pyright ignores just this pattern?

File: src/train/context_trainer.py [hunk 2]
```diff
@@ -73,16 +73,16 @@ def train(self, pbar: Optional[Any] = None) -> ContextModel:
             if self.checkpoint_freq > 0 and (i + self.step_offset) % self.checkpoint_freq == 0:
 
                 # save locally
-                local_dir_path = f"models/{os.path.basename(os.path.dirname(wandb.run.dir)).replace('run-', '')}"
+                local_dir_path = f"models/{os.path.basename(os.path.dirname(wandb.run.dir)).replace('run-', '')}" # pyright: ignore [reportOptionalMemberAccess]
```
Comment (nelson-lojo) [inline src/train/context_trainer.py:76]: See desired pyright ignore functionality in comment on `__main__.py`

File: .github/workflows/pyright.yml
Comment (nelson-lojo) [inline .github/workflows/pyright.yml:1]: The majority of the time in this run is installing the conda env (~70-100/120 seconds). We should do caching here if possible (my free actions minutes can only go so far lol)

**Review by nelson-lojo — COMMENTED** [review]
First glance over it looks good ... there are obvious ways to improve, but I think as long as we have something that works, it'll be a good starting point

File: .github/workflows/pyright.yml
Comment (nelson-lojo) [reply .github/workflows/pyright.yml:1]: Made an issue: #16

File: src/__main__.py (hunk 1 above)
Comment (nelson-lojo) [reply src/__main__.py:17]: What do you think @riensou ?

**Comment by nelson-lojo** [comment]
This PR is now codependent with #19 since type checking settings should be updated in the devcontainer file

File: src/__main__.py (hunk 1 above)
Comment (nelson-lojo) [reply src/__main__.py:18]: in-persons discussion agreed that this is acceptable, as it brings attention to the "need" for type stubbing (like a TODO), but is a reasonable middle ground for productivity now

**Comment by nelson-lojo** [comment]
There is a mismatch in pyright versions 1.1.357 (what I used to author this PR) and 1.1.356 (what the typechecking action runs). The entirety of the difference for us is captured in 8c1bd32. 

Ref'ing #19 to handle this later

**Review by nelson-lojo — COMMENTED** [review]
Total changes:
- Added pyright config
- Added pyright typechecking action to run on PR into main
- Added pyright:ignore directives at as fine a granularity as possible
- Changed root definition of FunctionClass.evaluate to only accept one type
- Changed root definition of ContextModel.forward to not accept keyword arguments (to encourage use of configuration through yaml)

**Review by riensou — APPROVED** [review]
lgtm

**Pull request merged**